src/network_event/client_event.rs

```diff
@@ -119,155 +119,219 @@ impl ClientEventAppExt for App {
-    channel: Res<ClientEventChannel<T>>,
-) {
-    for (client_id, message) in server.receive(*channel) {
-        match DefaultOptions::new().deserialize(&message) {
-            Ok(event) => {
-                trace!(
-                    "applying event `{}` from `{client_id:?}`",
-                    any::type_name::<T>()
-                );
-                client_events.send(FromClient { client_id, event });
-            }
-            Err(e) => debug!("unable to deserialize event from {client_id:?}: {e}"),
+fn map_and_send<T: Event + MapEntities + Serialize + Clone>(world: &mut World, channel: u8) {
+    world.resource_scope(|world, mut client: Mut<RepliconClient>| {
+        let entity_map = world.get_resource::<ServerEntityMap>().unwrap();
+        let events = world.get_resource::<Events<T>>().unwrap();
+
+        for mut event in events.get_reader().read(events).cloned() {
+            event.map_entities(&mut EventMapper(entity_map.to_server()));
+            let message = DefaultOptions::new()
+                .serialize(&event)
+                .expect("mapped client event should be serializable");
+
+            trace!("sending event `{}`", any::type_name::<T>());
+            client.send(channel, message);
         }
-    }
+    });
 }
 
-fn send<T: Event + Serialize>(
-    mut events: EventReader<T>,
-    mut client: ResMut<RepliconClient>,
-    channel: Res<ClientEventChannel<T>>,
-) {
-    for event in events.read() {
-        let message = DefaultOptions::new()
-            .serialize(&event)
-            .expect("client event should be serializable");
-
-        trace!("sending event `{}`", any::type_name::<T>());
-        client.send(*channel, message);
-    }
+struct ClientEventFns {
+    channel_id: u8,
+    send: SendFn,
+    resend_locally: fn(&mut World),
+    receive: ReceiveFn,
+    reset: fn(&mut World),
 }
 
-fn map_and_send<T: Event + MapEntities + Serialize + Clone>(
-    mut events: EventReader<T>,
-    mut client: ResMut<RepliconClient>,
-    entity_map: Res<ServerEntityMap>,
-    channel: Res<ClientEventChannel<T>>,
-) {
-    for mut event in events.read().cloned() {
-        event.map_entities(&mut EventMapper(entity_map.to_server()));
-        let message = DefaultOptions::new()
-            .serialize(&event)
-            .expect("mapped client event should be serializable");
-
-        trace!("sending event `{}`", any::type_name::<T>());
-        client.send(*channel, message);
+impl ClientEventFns {
+    fn new<T: Event + Serialize + DeserializeOwned>(
+        channel_id: u8,
+        send: SendFn,
+        receive: ReceiveFn,
+    ) -> Self {
+        Self {
+            channel_id,
+            send,
+            resend_locally: resend_locally::<T>,
+            receive,
+            reset: reset::<T>,
+        }
     }
 }
 
+type SendFn = fn(&mut World, u8);
+type ReceiveFn = fn(&mut World, u8);
+
+fn send<T: Event + Serialize>(world: &mut World, channel_id: u8) {
```

Comment:
I think we probably should make the API nicer by using two functions. First function will read the resource from the registry by ID (like you tried to do before), iterates over all events and passes it into a second function with the correct type.
And let user to override only the second function.